### PR TITLE
Add support for Results API

### PR DESCRIPTION
**Pull request by ro-oliveira95**
Add support for the ALFAsim's server Results API while also improving the way users can interact with the UQ results.

ASIM-5920

### CODE REVIEW

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 1]
```diff
@@ -149,28 +157,38 @@ class BaseUQMetaData:
...
+    property_name: str
+    element_name: str
+
+    def __str__(self):
+        return f"{self.property_name}@{self.element_name}"
+
+    @classmethod
+    def from_string(cls, raw_output_key: str) -> Self:
+        r = re.match(r"^(.*)@(.*)", raw_output_key)
+        assert r is not None, f"Invalid output key: {raw_output_key}"
```
Comment (nicoddemus) [inline src/alfasim_sdk/result_reader/aggregator.py:180]: This should be a normal exception (ValueError maybe?), not an assert.

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 2]
```diff
@@ -149,28 +157,38 @@ class BaseUQMetaData:
...
+    category: str
+    network_element_name: Optional[str]
+    position: Optional[float]
+    position_unit: Optional[str]
+    unit: str
 
 
-@attr.s(slots=True, hash=False)
+@dataclasses.dataclass(slots=True, frozen=True)
+class UPOutputKey:
```
Comment (nicoddemus) [inline src/alfasim_sdk/result_reader/aggregator.py:170]: Great abstraction!

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 3]
```diff
@@ -149,28 +157,38 @@ class BaseUQMetaData:
...
+
+    @classmethod
+    def from_string(cls, raw_output_key: str) -> Self:
+        r = re.match(r"^(.*)@(.*)", raw_output_key)
+        assert r is not None, f"Invalid output key: {raw_output_key}"
+        property_name, element_name = r.groups()
+        return cls(property_name, element_name)
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
```
Comment (nicoddemus) [inline src/alfasim_sdk/result_reader/aggregator.py:185]: This used `hash=False` previously... not sure why though, making it frozen seems reasonable.

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 4]
```diff
@@ -231,13 +243,30 @@ def map_data(
...
+    parametric_var_id: str
+    element_name: str
+
+    def __str__(self):
+        return f"{self.property_name}::{self.parametric_var_id}@{self.element_name}"
+
+    @classmethod
+    def from_string(cls, raw_output_key: str) -> Self:
+        r = re.match(r"^(.*)::(.*)@(.*)", raw_output_key)
+        assert r is not None, f"Invalid output key: {raw_output_key}"
```
Comment (nicoddemus) [inline src/alfasim_sdk/result_reader/aggregator.py:258]: Same about this being a proper exception.

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 5]
```diff
@@ -525,14 +616,18 @@ class ALFASimResultMetadata:
...
+    profiles: dict[OutputKeyType, ProfileMetaItem]
+    trends: dict[OutputKeyType, TrendMetaItem]
+    time_sets: list[SourceTimeSetKeyType]
+    time_sets_unit: str
+    time_steps_boundaries: tuple[
+        tuple[TimeStepIndex, TimeStepIndex], tuple[TimeStepIndex, TimeStepIndex]
+    ]
+    time_set_info: dict[
+        Literal["profiles", "trends"], dict[TimeStepIndex, TimeSetInfoItem]
+    ]  # TODO: verify if keys are TimeStep.
```
Comment (nicoddemus) [inline src/alfasim_sdk/result_reader/aggregator.py:628]: TODO

**Review by nicoddemus — COMMENTED** [review]
Great work!

File: tests/results/test_aggregator.py [hunk 6]
```diff
@@ -210,7 +209,9 @@ def test_concatenate_metadata_handle_time_sets(results: Results) -> None:
 
 def test_concatenate_metadata_error_conditions_directory(results: Results) -> None:
     md = results.metadata
-    other_md = attr.evolve(md, result_directory=md.result_directory / "other")
+    # other_md = attr.evolve(md, result_directory=md.result_directory / "other")
```
Comment (ro-oliveira95) [inline tests/results/test_aggregator.py:212]: - [ ] Remove left over

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 7]
```diff
@@ -204,22 +220,18 @@ def from_dict(cls, data: Dict[str, Any]) -> Self:
...
             )
 
-    items: Dict[str, UPItem] = attr.ib(validator=attr.validators.instance_of(Dict))
-    result_directory: Path = attr.ib(validator=attr.validators.instance_of(Path))
-
-    @classmethod
-    def empty(cls, result_directory: Path) -> Self:
-        return UncertaintyPropagationAnalysesMetaData(
-            items={}, result_directory=result_directory
-        )
```
Comment (ro-oliveira95) [inline src/alfasim_sdk/result_reader/aggregator.py:214]: A note for reviewers: this was removed there should be no case in which we have empty metadata for this type of results, once they don't have the concept of partial completion. The presence of these empty objects could lead to confusion and ultimately bugs, so they were ripped of.

File: src/alfasim_sdk/result_reader/aggregator.py (hunk 3 above)
Comment (ro-oliveira95) [reply src/alfasim_sdk/result_reader/aggregator.py:185]: Yeah I'm also not sure why, but at least for our internal use cases frozen works fine so I guess we're good.

File: src/alfasim_sdk/result_reader/aggregator.py (hunk 5 above)
Comment (ro-oliveira95) [reply src/alfasim_sdk/result_reader/aggregator.py:628]: - [ ] Remove TODO

I would ask @prusse-martin on this one but he is on vacation right now. I will leave this as `TimeStepIndex` for now but will ask him to do a post-mortem review on this and in the `Trend/ProfileMetaItem` definitions, then any changes can be addressed in a follow-up PR.

File: src/alfasim_sdk/result_reader/reader.py [hunk 8]
```diff
@@ -295,3 +330,154 @@ def get_profile_metadata(meta: Dict) -> ProfileMetadata:
             get_profile_metadata(profile_metadata)
             for profile_metadata in metadata.profiles.values()
         ]
+
+
+def _non_empty_dict_validator(values_type: type) -> Callable:
+    def validator(inst: Any, attribute: attr.Attribute, value: Any) -> None:
```
Comment (cauebs) [inline src/alfasim_sdk/result_reader/reader.py:336]: You've probably thought about this already, but it would be good if these had more specific types. Not sure if `object` instead of `Any` would work, but at least `Callable[[Any, attr.Attribute, Any] None]`?

Comment (cauebs) [reply src/alfasim_sdk/result_reader/reader.py:336]: If this is actionable, it applies to the other validators, of course.

**Review by cauebs — APPROVED** [review]

File: src/alfasim_sdk/result_reader/reader.py [hunk 9]
```diff
@@ -295,3 +330,154 @@ def get_profile_metadata(meta: Dict) -> ProfileMetadata:
...
+
+        return cls(
+            timeset=read_uq_time_set(
+                result_dir, GLOBAL_SENSITIVITY_ANALYSIS_GROUP_NAME
+            ),
+            coefficients=read_global_sensitivity_coefficients(metadata),
+            metadata=metadata,
+        )
+
+    def get_sensibility_curve(
```
Comment (BeneBr) [inline src/alfasim_sdk/result_reader/reader.py:382]: All this feature is called ```Global Sensitivity Analysis```, i believe that the correct name for this functions is:
```suggestion
    def get_sensitivity_curve(
```

File: tests/results/test_result_reader.py [hunk 10]
```diff
@@ -134,3 +152,185 @@ def test_metadata_string():
 
     global_trend = GlobalTrendMetadata("property")
     assert str(global_trend) == "property"
+
+
+def test_global_sensibility_analysis_results_reader(
```
Comment (BeneBr) [inline tests/results/test_result_reader.py:157]:
```suggestion
def test_global_sensitivity_analysis_results_reader(
```

**Review by BeneBr — APPROVED** [review]

File: src/alfasim_sdk/result_reader/reader.py (hunk 9 above)
Comment (ro-oliveira95) [reply src/alfasim_sdk/result_reader/reader.py:382]: Although "sensitivity" and "sensibility" convey the same meaning, I agree that it's better to be consistent. Will change that, good catch.

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 11]
```diff
@@ -93,19 +100,20 @@
...
-"""
 
-TimeSetInfoItem = namedtuple("TimeSetInfoItem", "global_start size uuid")
+class TimeSetInfoItem(NamedTuple):
+    global_start: TimeStepIndex
+    size: int
+    uuid: str
+
+
 TimeSetInfo = Dict[int, TimeSetInfoItem]
```
Comment (prusse-martin) [inline src/alfasim_sdk/result_reader/aggregator.py:117]: The key here could be `TimeStepIndex`.

File: src/alfasim_sdk/result_reader/aggregator.py [hunk 12]
```diff
@@ -923,7 +1020,7 @@ def read_time_set_info(
         if size > 0:
             limits[key] = TimeSetLimits(dataset[0], dataset[-1], dataset)
         time_set_info[key] = TimeSetInfoItem(
-            global_start=start, size=size, uuid=time_set_uuid
+            global_start=int(start), size=int(size), uuid=time_set_uuid
```
Comment (prusse-martin) [inline src/alfasim_sdk/result_reader/aggregator.py:1023]: Do you remember why did you need to type cast those arguments (start and size)?

Comment (ro-oliveira95) [reply src/alfasim_sdk/result_reader/aggregator.py:1023]: IIRC they were being assigned to numpy's type `np.uint8` and the (de)serialization engine (`pydantic` or `pyserde`, not sure which one) didn't considered it as a valid type for that particular field.
